pkg/operator/certupdater/cert_updater.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package certupdater contains an implementation of `tls.GetCertificate`.
+package certupdater
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io/fs"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/cert"
+)
+
+const tlsCAKey = "ca.crt"
+
+type certUpdater struct {
+	mu sync.RWMutex
+
+	logger logr.Logger
+
+	source      CertSource
+	currentCert *tls.Certificate
+	currentCA   *x509.Certificate
+
+	pollingInterval time.Duration
+}
+
+// Option configures a new certUpdater.
+type Option func(*certUpdater)
+
+// WithLogging provides a logger to certUpdater.
+func WithLogging(l logr.Logger) Option {
+	return func(cu *certUpdater) {
+		cu.logger = l.WithValues("package", "certupdater")
+	}
+}
+
+// WithPolling causes certUpdater to check for changes to certificates periodically.
+func WithPolling(d time.Duration) Option {
+	return func(cu *certUpdater) {
+		cu.pollingInterval = d
+	}
+}
+
+// New creates a new certUpdater.
+//
+//nolint:revive // Intentionally return unexported type, to use methods only.
+func New(source CertSource, opts ...Option) (*certUpdater, error) {
+	if source == nil {
+		return nil, fmt.Errorf("source must not be nil")
+	}
+	cu := &certUpdater{
+		source: source,
+	}
+	for _, opt := range opts {
+		opt(cu)
+	}
+	return cu, nil
+}
+
+// Start begins polling to periodically update certificates from the available sources.
+func (cu *certUpdater) Start(ctx context.Context) error {
+	go func() {
+		for {
+			select {
+			case <-time.After(cu.pollingInterval):
+				if err := cu.poll(ctx); err != nil {
+					cu.logger.Error(err, "Updating certs failed.")
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+func (cu *certUpdater) poll(ctx context.Context) error {
```

Comment:
suggestion: I may have brought this up before but why not have `poll` live within the source? The way a source watches for changes may be different. For example, the filesystem could use the native OS watch, flags will never poll, and if we ever add something like fetching Kubernetes secrets directly, that would be a Kubernetes Watch API.

Reply:
Basically, I found that polling is simpler and more reliable than other methods in Kubernetes.

Filesystem watches in Kubernetes behave oddly because of the way volumes work. When a change is detected using fsnotify, for instance, it gets reported as the file being removed. [Certwatcher gets around this by creating a new watch for each change](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.16.2/pkg/certwatcher/certwatcher.go#L182-L187). This is different from how fsnotify works outside Kubernetes. It's certainly possible to implement a similar solution, but it adds complexity.

Watches in Kubernetes are not free, and on my last team we started running into limitations on the Kubernetes API Server for how many watches could be set up at once, so I carried forward some bias from that. Each watch keeps a persistent connection open between the operator and the Kubernetes API Server. I also observed that Kubernetes secrets mounted as a volume (the intended use case) are cached and it takes about 2 minutes to propagate a change to a secret to the volume. So, whatever responsiveness you gain by using a watch is likely relatively inconsequential in that context.

Ultimately, using simple polling seemed like a better trade-off. It reduces complexity while having minimal impact on overall performance. That said, it would be possible to implement watches in the future, if we discovered a need to do so.